core/src/main/java/ru/spbau/blackout/serializationutils/EffectiveInputStream.java

```diff
@@ -0,0 +1,128 @@
+package ru.spbau.blackout.serializationutils;
+
+import com.badlogic.gdx.math.Vector2;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+
+public class EffectiveInputStream extends InputStream {
+
+    private final InputStream in;
+
+    public EffectiveInputStream(InputStream in) {
+        this.in = in;
+    }
+
+    @Override
+    public int read() throws IOException {
+        return in.read();
+    }
+
+    public synchronized boolean readBoolean() throws IOException {
+        return read() == 1;
+    }
+
+    public synchronized byte readByte() throws IOException {
+        return (byte) read();
+    }
+
+    public synchronized char readChar() throws IOException {
+        final int size = Character.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getChar();
+    }
+
+    public synchronized short readShort() throws IOException {
+        final int size = Short.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getShort();
+    }
+
+    public synchronized float readFloat() throws IOException {
+        final int size = Float.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getFloat();
+    }
+
+    public synchronized double readDouble() throws IOException {
+        final int size = Double.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getDouble();
+    }
+
+    public synchronized int readInt() throws IOException {
+        final int size = Integer.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getInt();
+    }
+
+    public synchronized long readLong() throws IOException {
+        final int size = Long.SIZE / Byte.SIZE;
+        byte[] bytes = new byte[size];
+        if (read(bytes) != size) {
+            throw new IOException();
+        }
+        return ByteBuffer.wrap(bytes).getLong();
+    }
+
+    public synchronized String readString() throws IOException {
+        boolean isPresent = readBoolean();
+        if (isPresent) {
+            int size = readInt();
+            byte[] bytes = new byte[size];
+            if (read(bytes) != size) {
```

Comment:
See this:

http://stackoverflow.com/questions/12717756/strange-byte-behavior-reading-from-a-url